MathLib/LinAlg/MatrixVectorTraits.cpp

```diff
@@ -0,0 +1,166 @@
+/**
+ * \copyright
+ * Copyright (c) 2012-2016, OpenGeoSys Community (http://www.opengeosys.org)
+ *            Distributed under a Modified BSD License.
+ *              See accompanying file LICENSE.txt or
+ *              http://www.opengeosys.org/project/license
+ *
+ */
+
+#include "MatrixVectorTraits.h"
+
+#ifdef OGS_USE_EIGEN
+
+namespace MathLib
+{
+
+std::unique_ptr<Eigen::MatrixXd>
+MatrixVectorTraits<Eigen::MatrixXd>::
+newInstance()
+{
+    return std::unique_ptr<Eigen::MatrixXd>(new Eigen::MatrixXd);
+}
+
+std::unique_ptr<Eigen::MatrixXd>
+MatrixVectorTraits<Eigen::MatrixXd>::
+newInstance(Eigen::MatrixXd const& A)
+{
+    return std::unique_ptr<Eigen::MatrixXd>(new Eigen::MatrixXd(A));
+}
+
+std::unique_ptr<Eigen::MatrixXd>
+MatrixVectorTraits<Eigen::MatrixXd>::
+newInstance(MatrixSpecifications const& spec)
+{
+    return std::unique_ptr<Eigen::MatrixXd>(new Eigen::MatrixXd(spec.nrows, spec.ncols));
+}
+
+std::unique_ptr<Eigen::VectorXd>
+MatrixVectorTraits<Eigen::VectorXd>::
+newInstance()
+{
+    return std::unique_ptr<Eigen::VectorXd>(new Eigen::VectorXd);
+}
+
+std::unique_ptr<Eigen::VectorXd>
+MatrixVectorTraits<Eigen::VectorXd>::
+newInstance(Eigen::VectorXd const& A)
+{
+    return std::unique_ptr<Eigen::VectorXd>(new Eigen::VectorXd(A));
+}
+
+std::unique_ptr<Eigen::VectorXd>
+MatrixVectorTraits<Eigen::VectorXd>::
+newInstance(MatrixSpecifications const& spec)
+{
+    return std::unique_ptr<Eigen::VectorXd>(new Eigen::VectorXd(spec.nrows));
+}
+
+} // namespace MathLib
+
+#endif // OGS_USE_EIGEN
+
+
+#ifdef USE_PETSC
+
+namespace MathLib
+{
+
+std::unique_ptr<PETScMatrix>
+MatrixVectorTraits<PETScMatrix>::
+newInstance()
+{
+    return std::unique_ptr<PETScMatrix>(new PETScMatrix(0, 0)); // TODO default constructor
+}
+
+std::unique_ptr<PETScMatrix>
+MatrixVectorTraits<PETScMatrix>::
+newInstance(PETScMatrix const& A)
+{
+    return std::unique_ptr<PETScMatrix>(new PETScMatrix(A));
+}
+
+std::unique_ptr<PETScMatrix>
+MatrixVectorTraits<PETScMatrix>::
+newInstance(MatrixSpecifications const& spec)
+{
+    return std::unique_ptr<PETScMatrix>(new PETScMatrix(spec.nrows)); // TODO sparsity pattern
+}
+
+std::unique_ptr<PETScVector>
+MatrixVectorTraits<PETScVector>::
+newInstance()
+{
+    return std::unique_ptr<PETScVector>(new PETScVector);
+}
+
+std::unique_ptr<PETScVector>
+MatrixVectorTraits<PETScVector>::
+newInstance(PETScVector const& x)
+{
+    return std::unique_ptr<PETScVector>(new PETScVector(x));
+}
+
+std::unique_ptr<PETScVector>
+MatrixVectorTraits<PETScVector>::
+newInstance(MatrixSpecifications const& spec)
+{
+    return std::unique_ptr<PETScVector>(new PETScVector(spec.nrows));
+}
+
+} // namespace MathLib
+
+
+#elif defined(OGS_USE_EIGEN)
```

Comment:
The first ifdef is for dense Eigen matrices. That case will be always built when we use the Eigen library (that's de facto always).
The second ifdef, or series of ifdefs (for PETSc or Eigen) resembles the ifdefs in NumericsConfig.h, i.e., only those methods that work with the global matrix and vector types are enabled.
Thus, there is an advantage from merging those sections (fewer ifdefs), but there's also a disadvantage: EigenMatrix case is always compiled. Hence, I'd like to keep the status quo.